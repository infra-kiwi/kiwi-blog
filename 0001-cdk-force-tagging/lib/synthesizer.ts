import * as cdk from 'aws-cdk-lib';
import simpleGit from 'simple-git';

export type KiwiSynthesizerProps = cdk.DefaultStackSynthesizerProps & {
  forcedTags: Record<string, string>;
};

export class KiwiSynthesizer extends cdk.DefaultStackSynthesizer {
  readonly forcedTags: KiwiSynthesizerProps['forcedTags'];

  constructor(props: KiwiSynthesizerProps) {
    super(props);
    this.forcedTags = props.forcedTags;
  }

  bind(stack: cdk.Stack) {
    super.bind(stack);

    // Add every forced tag to the Stack
    for (const tag in this.forcedTags) {
      cdk.Tags.of(this.boundStack).add(tag, this.forcedTags[tag]);
    }
  }
}

export async function getSynthesizer(synthProps?: cdk.DefaultStackSynthesizerProps) {
  // Extract the current repository name
  const remoteUrl = (await simpleGit().remote(['get-url', 'origin']))!.trim();
  const repository = /github.com[:/]([^/]+\/[^/]+)\.git/.exec(remoteUrl)![1];

  return new KiwiSynthesizer({
    ...synthProps,
    forcedTags: {
      Repository: repository
    }
  });
}
