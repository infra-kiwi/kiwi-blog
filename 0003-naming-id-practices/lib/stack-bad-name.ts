/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { NotificationGroup } from './constructs-bad-name/notification-group';
import { SNSTopicWrapper } from './constructs-bad-name/sns-topic-wrapper';

export class StackBadName extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    new SNSTopicWrapper(this, 'NotificationsFunFacts');

    new NotificationGroup(this, 'Default');
    new NotificationGroup(this, 'TeamInfra');
  }
}
