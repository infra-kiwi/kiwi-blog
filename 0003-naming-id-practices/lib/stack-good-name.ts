import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { NotificationGroup } from './constructs-good-name/notification-group';
import { SNSTopicWrapper } from './constructs-good-name/sns-topic-wrapper';

export class StackGoodName extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    new SNSTopicWrapper(this, 'NotificationsFunFacts');

    new NotificationGroup(this, 'Default');
    new NotificationGroup(this, 'TeamInfra');
  }
}
