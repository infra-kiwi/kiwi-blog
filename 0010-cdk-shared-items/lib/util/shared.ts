import { aws_events } from 'aws-cdk-lib';

/**
 * SharedItems is a global struct that contains items shared
 * among multiple Stacks.
 *
 * To set and get items, use the `setSharedItem`/`getSharedItem` functions.
 */
export interface SharedItems {
  // eventBusShared is the core EventBus of the application
  eventBusShared: aws_events.EventBus;
}

// Initialize the struct as an empty object
const sharedItems: SharedItems = {} as never;

// setSharedItem sets an item in the SharedItems struct, and
// enforces type constraints (valid key, valid value)
export function setSharedItem<Key extends keyof SharedItems>(
  key: Key,
  value: SharedItems[Key]
) {
  sharedItems[key] = value;
}

// getSharedItem gets an item from the SharedItems struct, and
// enforces type constraints (valid key)
export function getSharedItem<Key extends keyof SharedItems>(
  key: Key
): SharedItems[Key] {
  const value = sharedItems[key];
  if (value === undefined) {
    throw new Error(`Tried to get a shared item before setting it: ${key}`);
  }
  return value;
}
