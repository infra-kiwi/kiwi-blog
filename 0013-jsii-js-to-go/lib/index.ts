/*
 * Copyright (c) 2025. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

export class Constants {
  static readonly name = 'Mario';
}

export interface Home {
  readonly address: string;
}

export class Addressor {
  getFullGreeting(name: string, home: Home) {
    return `Hello ${Constants.name}! You live at ${home.address}.`;
  }
}
