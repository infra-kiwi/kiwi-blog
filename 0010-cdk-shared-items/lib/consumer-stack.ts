/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from 'aws-cdk-lib';
import {
  aws_events,
  aws_events_targets,
  aws_lambda,
  aws_lambda_nodejs,
  CfnOutput,
  Stack
} from 'aws-cdk-lib';
import { Construct } from 'constructs';
import path from 'node:path';
import { getSharedItem } from './util/shared';
import { LoggingFormat } from 'aws-cdk-lib/aws-lambda';

export class ConsumerStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // https://medium.com/infrakiwi/aws-cdk-the-easy-node-js-lambda-ec4dc0d60807
    const lambda = new aws_lambda_nodejs.NodejsFunction(
      this,
      'ConsumerLambda',
      {
        functionName: `${Stack.of(this).stackName}-lambda`,
        entry: path.join(__dirname, './lambda/consumer.ts'),
        runtime: aws_lambda.Runtime.NODEJS_20_X,
        handler: 'index.handler',
        bundling: {
          externalModules: ['@aws-sdk/*']
        },
        loggingFormat: LoggingFormat.JSON
      }
    );
    new CfnOutput(this, 'ConsumerLambdaName', {
      value: lambda.functionName
    });
    new CfnOutput(this, 'ConsumerLambdaLogGroupName', {
      value: lambda.logGroup.logGroupName
    });

    // Define a rule that forwards all events created by the ProducerLambda
    // to the ConsumerLambda
    const rule = new aws_events.Rule(this, 'RuleConsumer', {
      // Use the shared EventBus
      eventBus: getSharedItem('eventBusShared'),
      eventPattern: {
        source: ['ProducerLambda']
      }
    });
    rule.addTarget(new aws_events_targets.LambdaFunction(lambda));
  }
}
