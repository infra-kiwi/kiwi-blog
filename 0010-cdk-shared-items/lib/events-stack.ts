import * as cdk from 'aws-cdk-lib';
import { aws_events } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { setSharedItem } from './util/shared';

export class EventsStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const eventBus = new aws_events.EventBus(this, 'EventBusShared');

    // Set the EventBus as a global element
    setSharedItem('eventBusShared', eventBus);
  }
}
