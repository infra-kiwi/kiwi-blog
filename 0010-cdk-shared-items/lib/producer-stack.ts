/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from 'aws-cdk-lib';
import { aws_lambda, aws_lambda_nodejs, CfnOutput, Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import path from 'node:path';
import { getSharedItem } from './util/shared';

/*
Defines a Lambda function that generates an EventBridge event on demand
 */
export class ProducerStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Use the shared EventBus as a destination for new events
    const eventBus = getSharedItem('eventBusShared');

    // https://medium.com/infrakiwi/aws-cdk-the-easy-node-js-lambda-ec4dc0d60807
    const lambda = new aws_lambda_nodejs.NodejsFunction(
      this,
      'ProducerLambda',
      {
        functionName: `${Stack.of(this).stackName}-lambda`,
        entry: path.join(__dirname, './lambda/producer.ts'),
        runtime: aws_lambda.Runtime.NODEJS_20_X,
        handler: 'index.handler',
        bundling: {
          externalModules: ['@aws-sdk/*']
        },
        environment: {
          // Pass the EventBus name to the function, so that
          // the function can put events to it
          EVENT_BUS_NAME: eventBus.eventBusName
        }
      }
    );
    new CfnOutput(this, 'ProducerLambdaName', {
      value: lambda.functionName
    });

    // Add the right IAM perms for the Lambda function to put events
    // to the core EventBus
    eventBus.grantPutEventsTo(lambda);
  }
}
