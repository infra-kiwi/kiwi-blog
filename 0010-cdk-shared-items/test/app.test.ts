import { ProducerStack } from '../lib/producer-stack';
import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { getSharedItem } from '../lib/util/shared';
import { EventsStack } from '../lib/events-stack';
import { ConsumerStack } from '../lib/consumer-stack';

test('SharedItems utility', async () => {
  const app = new App();

  // Getting a shared item before it has been set should fail
  expect(() => getSharedItem('eventBusShared')).toThrow(
    'Tried to get a shared item before setting it: eventBusShared'
  );

  // Create the EventStack, which sets our shared item
  const eventsStack = new EventsStack(app, 'Events');
  expect(getSharedItem('eventBusShared')).not.toBeUndefined();

  // Init the other stacks
  const producerStack = new ProducerStack(app, 'Producer');
  const consumerStack = new ConsumerStack(app, 'Consumer');

  {
    const template = Template.fromStack(eventsStack);
    expect(template).toMatchSnapshot();
  }
  {
    const template = Template.fromStack(producerStack);
    expect(template).toMatchSnapshot();
  }
  {
    const template = Template.fromStack(consumerStack);
    expect(template).toMatchSnapshot();
  }
});
