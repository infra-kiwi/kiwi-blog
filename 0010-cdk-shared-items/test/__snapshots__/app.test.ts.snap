// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SharedItems utility 1`] = `
{
  "Outputs": {
    "ExportsOutputFnGetAttEventBusSharedF95CAE4FArn37C7766A": {
      "Export": {
        "Name": "Events:ExportsOutputFnGetAttEventBusSharedF95CAE4FArn37C7766A",
      },
      "Value": {
        "Fn::GetAtt": [
          "EventBusSharedF95CAE4F",
          "Arn",
        ],
      },
    },
    "ExportsOutputRefEventBusSharedF95CAE4F55165ED6": {
      "Export": {
        "Name": "Events:ExportsOutputRefEventBusSharedF95CAE4F55165ED6",
      },
      "Value": {
        "Ref": "EventBusSharedF95CAE4F",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "EventBusSharedF95CAE4F": {
      "Properties": {
        "Name": "EventsEventBusSharedCBE568B1",
      },
      "Type": "AWS::Events::EventBus",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`SharedItems utility 2`] = `
{
  "Outputs": {
    "ProducerLambdaName": {
      "Value": {
        "Ref": "ProducerLambdaC55AC1D4",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ProducerLambdaC55AC1D4": {
      "DependsOn": [
        "ProducerLambdaServiceRoleDefaultPolicy8D34AFD0",
        "ProducerLambdaServiceRole37E53FC3",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "b4d3267bd8cbb97e96f18ec628836307f590f951a57a645ac221ff1a3a0d4937.zip",
        },
        "Environment": {
          "Variables": {
            "EVENT_BUS_NAME": {
              "Fn::ImportValue": "Events:ExportsOutputRefEventBusSharedF95CAE4F55165ED6",
            },
          },
        },
        "FunctionName": "Producer-lambda",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "ProducerLambdaServiceRole37E53FC3",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ProducerLambdaServiceRole37E53FC3": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ProducerLambdaServiceRoleDefaultPolicy8D34AFD0": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "events:PutEvents",
              "Effect": "Allow",
              "Resource": {
                "Fn::ImportValue": "Events:ExportsOutputFnGetAttEventBusSharedF95CAE4FArn37C7766A",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ProducerLambdaServiceRoleDefaultPolicy8D34AFD0",
        "Roles": [
          {
            "Ref": "ProducerLambdaServiceRole37E53FC3",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`SharedItems utility 3`] = `
{
  "Mappings": {
    "LatestNodeRuntimeMap": {
      "af-south-1": {
        "value": "nodejs20.x",
      },
      "ap-east-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-2": {
        "value": "nodejs20.x",
      },
      "ap-northeast-3": {
        "value": "nodejs20.x",
      },
      "ap-south-1": {
        "value": "nodejs20.x",
      },
      "ap-south-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-1": {
        "value": "nodejs20.x",
      },
      "ap-southeast-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-3": {
        "value": "nodejs20.x",
      },
      "ap-southeast-4": {
        "value": "nodejs20.x",
      },
      "ap-southeast-5": {
        "value": "nodejs20.x",
      },
      "ap-southeast-7": {
        "value": "nodejs20.x",
      },
      "ca-central-1": {
        "value": "nodejs20.x",
      },
      "ca-west-1": {
        "value": "nodejs20.x",
      },
      "cn-north-1": {
        "value": "nodejs20.x",
      },
      "cn-northwest-1": {
        "value": "nodejs20.x",
      },
      "eu-central-1": {
        "value": "nodejs20.x",
      },
      "eu-central-2": {
        "value": "nodejs20.x",
      },
      "eu-isoe-west-1": {
        "value": "nodejs18.x",
      },
      "eu-north-1": {
        "value": "nodejs20.x",
      },
      "eu-south-1": {
        "value": "nodejs20.x",
      },
      "eu-south-2": {
        "value": "nodejs20.x",
      },
      "eu-west-1": {
        "value": "nodejs20.x",
      },
      "eu-west-2": {
        "value": "nodejs20.x",
      },
      "eu-west-3": {
        "value": "nodejs20.x",
      },
      "il-central-1": {
        "value": "nodejs20.x",
      },
      "me-central-1": {
        "value": "nodejs20.x",
      },
      "me-south-1": {
        "value": "nodejs20.x",
      },
      "mx-central-1": {
        "value": "nodejs20.x",
      },
      "sa-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-2": {
        "value": "nodejs20.x",
      },
      "us-gov-east-1": {
        "value": "nodejs20.x",
      },
      "us-gov-west-1": {
        "value": "nodejs20.x",
      },
      "us-iso-east-1": {
        "value": "nodejs18.x",
      },
      "us-iso-west-1": {
        "value": "nodejs18.x",
      },
      "us-isob-east-1": {
        "value": "nodejs18.x",
      },
      "us-west-1": {
        "value": "nodejs20.x",
      },
      "us-west-2": {
        "value": "nodejs20.x",
      },
    },
  },
  "Outputs": {
    "ConsumerLambdaLogGroupName": {
      "Value": {
        "Fn::GetAtt": [
          "ConsumerLambdaLogRetention18BBBCF6",
          "LogGroupName",
        ],
      },
    },
    "ConsumerLambdaName": {
      "Value": {
        "Ref": "ConsumerLambdaF347BC65",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ConsumerLambdaF347BC65": {
      "DependsOn": [
        "ConsumerLambdaServiceRole9D55AA81",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "322dff04884d4eaf9f26bdcc052ebeeb335d3515721cc0de4a4a63db7f835e8f.zip",
        },
        "FunctionName": "Consumer-lambda",
        "Handler": "index.handler",
        "LoggingConfig": {
          "LogFormat": "JSON",
        },
        "Role": {
          "Fn::GetAtt": [
            "ConsumerLambdaServiceRole9D55AA81",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ConsumerLambdaLogRetention18BBBCF6": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "ConsumerLambdaF347BC65",
              },
            ],
          ],
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "ConsumerLambdaServiceRole9D55AA81": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "2819175352ad1ce0dae768e83fc328fb70fb5f10b4a8ff0ccbcb791f02b0716d.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": {
          "Fn::FindInMap": [
            "LatestNodeRuntimeMap",
            {
              "Ref": "AWS::Region",
            },
            "value",
          ],
        },
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "RuleConsumer520E0E13": {
      "Properties": {
        "EventBusName": {
          "Fn::ImportValue": "Events:ExportsOutputRefEventBusSharedF95CAE4F55165ED6",
        },
        "EventPattern": {
          "source": [
            "ProducerLambda",
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "ConsumerLambdaF347BC65",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "RuleConsumerAllowEventRuleConsumerConsumerLambdaB2105ABA35D03119": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ConsumerLambdaF347BC65",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "RuleConsumer520E0E13",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
